### PR TITLE
fix test when nested jpaDerivedIdentifier is used

### LINKED ISSUE #17906: Integration tests fail with nested jpaDerivedIdentifier
##### **Overview of the issue**
Integration tests fail when I define nested entities using jpaDerivedIdentifier
the integration tests fail for entities at or lower than grandchild level

Given the following entities:  
  - Animal
    - Fish
    - Bird
    - Mammal
      - Marsupial
      - Rodent
      - Cat
        - DomesticatedCat

And the following relationships:
relationship OneToOne {
  Fish{animal required} to Animal with jpaDerivedIdentifier
  Bird{animal required} to Animal with jpaDerivedIdentifier
  Mammal{animal required} to Animal with jpaDerivedIdentifier
    Marsupial{mammal required} to Mammal with jpaDerivedIdentifier
    Rodent{mammal required} to Mammal with jpaDerivedIdentifier
    Cat{mammal required} to Mammal with jpaDerivedIdentifier
      DomesticatedCat{cat required} to Cat with jpaDerivedIdentifier
}

generate the app
> jhipster jdl tree.jdl

test the app 
> mvn verify

The following errors are reported:

> [ERROR] Errors:
> [ERROR]   CatResourceIT.updateCatMapsIdAssociationWithNewId:150 » EntityExists A differe...
> [ERROR]   DomesticatedCatResourceIT.updateDomesticatedCatMapsIdAssociationWithNewId:155 » EntityExists
> [ERROR]   MarsupialResourceIT.updateMarsupialMapsIdAssociationWithNewId:150 » EntityExists
> [ERROR]   RodentResourceIT.updateRodentMapsIdAssociationWithNewId:150 » EntityExists A d...
> 

The tests for AnimalResourceIT, FishResourceIT, BirdResourceIT and MammalResourceIT all work,  
but anything lower than one level has a problem

##### **Motivation for or Use Case**

This isn't a problem for the app  at runtime, it seems to be an integration test issue only.

##### **Reproduce the error**

See the Overview, jdl will be provided.

##### **JHipster Version(s)**

7.5.0

##### **JHipster configuration**

INFO! Using JHipster version installed locally in current project's node_modules
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
tree@0.0.1-SNAPSHOT C:\github\jhip-treebug
`-- generator-jhipster@7.5.0

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "clientFramework": "angularX",
    "serverPort": "8080",
    "serviceDiscoveryType": false,
    "skipUserManagement": false,
    "withAdminUi": true,
    "baseName": "tree",
    "buildTool": "maven",
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": ["en"],
    "messageBroker": false,
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "testFrameworks": [],
    "websocket": false,
    "enableGradleEnterprise": false,
    "gradleEnterpriseHost": "",
    "applicationType": "monolith",
    "packageName": "com.jhipster.demo.tree",
    "packageFolder": "com/jhipster/demo/tree",
    "jhipsterVersion": "7.5.0",
    "clientPackageManager": "npm",
    "dtoSuffix": "DTO",
    "entitySuffix": "",
    "reactive": false,
    "clientTheme": "none",
    "clientThemeVariant": "",
    "applicationIndex": 0,
    "entities": ["Animal", "Fish", "Bird", "Mammal", "Marsupial", "Rodent", "Cat", "DomesticatedCat"],
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "blueprints": [],
    "otherModules": [],
    "pages": [],
    "nativeLanguage": "en",
    "creationTimestamp": 1645088543533,
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",    "devServerPort": 4200,
    "lastLiquibaseTimestamp": 1645089023000
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Animal {
  name String required minlength(3)
  description String
}
entity Fish {
  hasScales Boolean required
}
entity Bird {
  wakingBehaviour WakingBehaviour required
}
entity Mammal {
  numberOfLegs Integer required min(2)
  numberOfArms Integer required min(0)
}
entity Marsupial {
  isEndangered Boolean required
}
entity Rodent {
  hasTail Boolean required
}
entity Cat {
  livesLeft Integer required
}
entity DomesticatedCat {
  hairType HairType required
}
enum WakingBehaviour {
  DIURNAL,
  NOCTURNAL
}
enum HairType {
  HAIRLESS,
  LONG,
  SHORT
}

relationship OneToOne {
  Fish{animal required} to Animal
  Bird{animal required} to Animal
  Mammal{animal required} to Animal
  Marsupial{mammal required} to Mammal
  Rodent{mammal required} to Mammal
  Cat{mammal required} to Mammal
  DomesticatedCat{cat required} to Cat
}

paginate Animal, Fish, Bird, Mammal, Marsupial, Rodent with infinite-scroll

</pre>
</details>

##### **Environment and Tools**

openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment Temurin-11.0.13+8 (build 11.0.13+8)
OpenJDK 64-Bit Server VM Temurin-11.0.13+8 (build 11.0.13+8, mixed mode)

git version 2.34.1.windows.1

node: v16.13.2

npm: 8.1.2

Docker version 20.10.12, build e91ed57

Docker Compose version v2.2.3

No change to package.json was detected. No package manager install will be executed.
Congratulations, JHipster execution is complete!

##### **Browsers and Operating System**

Windows 10 Pro
Chrome, but this issue is not browser related

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by emilpaw**
when a relationship is used with the jpaDerivedIdentifier option and the
related entity also has a relationship with that option set the
`updateEntityMapsIdAssociationWithNewId` test fails. This
is caused by the `createUpdatedEntity` reusing the needed parent
entity. This can't work because a parent entity, which already has a child,
can't be used as the parent for a new entity. The reason being
jpaDerivedIdentifier enforces that parent and child share the same id,
which can't be fulfilled with a reused parent entity.
Remove the reuse of the parent entity in `createUpdatedEntity` when the
entity uses jpaDerivedIdentifier to fix the test.

Fixes #17906

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by emilpaw** [comment]
The failing test `react-gradle-mysql-es-noi18n-mapsid` has nothing to do with my change. I wonder why this test hasn't failed before since I did not introduce the issue. 

Either way, I'd like to add a test for the changes I made but I haven't figured out where exactly it should be placed. Could someone help me with that?

**Comment by mraible** [comment]
Tests pass. Is this good enough?

**Comment by emilpaw** [comment]
@mraible The existing tests never caught the issue this PR fixes. As a measure to avoid regressions, in my opinion, we should add one JDL integration test. I am just still not sure how to go about that.

The JDL I want to test should look something like that:
```
entity Animal
entity Mamal
entity Cat

relationship OneToOne {
    Mamal to Animal with jpaDerivedIdentifier
    Cat to Mamal with jpaDerivedIdentifier
}
```
Where should I add such a test case?

**Comment by mshima** [comment]
It's possible to add an `MapsIdGrandChildEntity*DTO` json referring to https://github.com/jhipster/generator-jhipster/blob/2aa6f6d6581f98a8edb08379b252bc20e724b212/test-integration/samples/.jhipster/MapsIdChildEntityWithoutDTO.json

Or add to one of customId samples:
https://github.com/jhipster/generator-jhipster/blob/2aa6f6d6581f98a8edb08379b252bc20e724b212/test-integration/samples/jdl-entities/custom-id.jdl#L19
https://github.com/jhipster/generator-jhipster/blob/2aa6f6d6581f98a8edb08379b252bc20e724b212/test-integration/samples/jdl-entities/custom-id.jdl#L52
https://github.com/jhipster/generator-jhipster/blob/2aa6f6d6581f98a8edb08379b252bc20e724b212/test-integration/samples/jdl-entities/custom-id.jdl#L87
https://github.com/jhipster/generator-jhipster/blob/2aa6f6d6581f98a8edb08379b252bc20e724b212/test-integration/samples/jdl-entities/custom-id.jdl#L120

**Comment by emilpaw** [comment]
@mshima Thank you. 

I added the test cases.

**Comment by mshima** [comment]
@emilpaw is this PR ready?

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 1]
```diff
@@ -517,13 +517,17 @@ if (field.fieldTypeString || field.blobContentTypeText) {
         <%_ } else { _%>
         <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %>;
           <%_ if (databaseTypeSql && !reactive) { _%>
+            <%_ if (!(isUsingMapsId && fieldStatus === "UPDATED_")) { _%>
```
Comment (mshima) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:520]:
```suggestion
            <%_ if (!isUsingMapsId || fieldStatus !== "UPDATED_")) { _%>
```

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 2]
```diff
@@ -517,13 +517,17 @@ if (field.fieldTypeString || field.blobContentTypeText) {
...
         if (TestUtil.findAll(em, <%= asEntity(otherEntityNameCapitalized) %>.class).isEmpty()) {
+            <%_ } _%>
             <%= otherEntityName %> = <%= createEntityPrefix %><%= otherEntityNameCapitalized %>ResourceIT.create<% if (fieldStatus === 'UPDATED_') { %>Updated<% } %>Entity(em)<%= createEntityPostfix %>;
             em.persist(<%= otherEntityName %>);
             em.flush();
+            <%_ if (!(isUsingMapsId && fieldStatus === "UPDATED_")) { _%>
```
Comment (mshima) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:526]:
```suggestion
            <%_ if (!isUsingMapsId || fieldStatus !== "UPDATED_") { _%>
```

**Review by mshima — COMMENTED** [review]
Simplify the condition.

**Review by mshima — APPROVED** [review]

File: test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
+  "fluentMethods": true,
+  "jpaMetamodelFiltering": false,
+  "pagination": "no",
+  "relationships": [
+    {
+      "id": true,
+      "otherEntityField": "id",
+      "otherEntityName": "mapsIdChildEntityWithDTO",
+      "otherEntityRelationshipName": "mapsIdGrandchildEntityWithDTO",
```
Comment (mshima) [inline test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json:16]:
```suggestion
```

File: test-integration/samples/.jhipster/MapsIdGrandchildEntityWithDTO.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
+  "fluentMethods": true,
+  "jpaMetamodelFiltering": false,
+  "pagination": "no",
+  "relationships": [
+    {
+      "id": true,
+      "otherEntityField": "id",
+      "otherEntityName": "mapsIdChildEntityWithDTO",
+      "otherEntityRelationshipName": "mapsIdGrandchildEntityWithDTO",
```
Comment (mshima) [inline test-integration/samples/.jhipster/MapsIdGrandchildEntityWithDTO.json:16]:
```suggestion
```

**Review by mshima — COMMENTED** [review]
Remove back reference, it's not there

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 5]
```diff
@@ -517,13 +517,17 @@ if (field.fieldTypeString || field.blobContentTypeText) {
         <%_ } else { _%>
         <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %>;
           <%_ if (databaseTypeSql && !reactive) { _%>
+            <%_ if (!isUsingMapsId || fieldStatus !== "UPDATED_")) { _%>
```
Comment (mshima) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:520]:
```suggestion
            <%_ if (!isUsingMapsId || fieldStatus !== "UPDATED_") { _%>
```

File: test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json (hunk 3 above)
Comment (emilpaw) [reply test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json]: `"otherEntityRelationshipName"` is needed even if the relationship is unidirectional, right?

File: test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json [hunk 6]
```diff
@@ -12,9 +12,10 @@
     {
       "id": true,
       "otherEntityField": "id",
-      "otherEntityName": "mapsIdChildEntityWithDTO",
+      "otherEntityName": "mapsIdChildEntityWithoutDTO",
+      "otherEntityRelationshipName": "mapsIdGrandchildEntityWithoutDTO",
       "ownerSide": true,
-      "relationshipName": "mapsIdChildEntityWithDTO",
```
Comment (emilpaw) [inline test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json]: Good catch. Thanks.

**Comment by emilpaw** [comment]
I squashed the commits and now it's ready to merge.

**Review by mshima — COMMENTED** [review]
We need investigation if one-to-one works without otherEntity filtering (removes already used ids). But not really related to this PR.

For now, if `otherEntityRelationshipName` is set, otherEntity is not set and the relationship is id, we should throw. Filtering will not work as expected.

**Comment by emilpaw** [comment]
@mshima Done.

**Review by mshima — APPROVED** [review]
LGTM

**Comment by mshima** [comment]
Thanks @emilpaw

**Comment by emilpaw** [comment]
Bug bounty claimed: https://opencollective.com/generator-jhipster/expenses/115030 Thanks :)

**Comment by DanielFran** [comment]
@emilpaw approved

**Pull request merged**